pontoon/sync/tasks.py

```diff
@@ -97,62 +92,61 @@ def sync_project(
 
     log.info("Syncing project {0}.".format(db_project.slug))
 
-    # Sync source strings. We cannot sync sources if locale is specified,
-    # because that would apply any source string changes to the specified locale only.
-    if locale:
-        source_changes = {}
-    else:
-        source_changes = sync_sources(db_project, now, force, no_pull)
-        # Skip syncing translations if we already know there's nothing to sync
-        if not source_changes:
-            project_sync_log.skip()
-            return
+    source_changes = sync_sources(db_project, now, force, no_pull)
+    # Skip syncing translations if no source directory found
+    if not source_changes:
+        project_sync_log.skip()
+        return
 
     # Sync translations
     sync_translations(
-        project_pk,
-        project_sync_log.pk,
+        db_project,
+        project_sync_log,
         now,
+        source_changes.get("source_changed"),
         source_changes.get("added_paths"),
         source_changes.get("removed_paths"),
         source_changes.get("changed_paths"),
         source_changes.get("new_entities"),
-        locale=locale,
         no_pull=no_pull,
         no_commit=no_commit,
-        full_scan=force,
+        force=force,
     )
 
 
 def sync_sources(db_project, now, force, no_pull):
-    # Pull source repository
+    # Pull from source repository
     if no_pull:
-        source_repo_changed = True  # Assume changed
+        source_changed = True
     else:
-        source_repo_changed, _ = pull_changes(db_project)
+        log.info(
+            "Pulling source changes for project {0} started.".format(db_project.slug)
+        )
+        source_changed = pull_source_repo_changes(db_project)
+        log.info(
+            "Pulling source changes for project {0} complete.".format(db_project.slug)
+        )
 
-    # If the only repo hasn't changed since the last sync and there are
+    # If the repos haven't changed since the last sync and there are
     # no Pontoon-side changes for this project, quit early.
     if (
         not force
         and not db_project.needs_sync
-        and not source_repo_changed
+        and not source_changed
         and db_project.has_single_repo
     ):
         log.info("Skipping project {0}, no changes detected.".format(db_project.slug))
         return False
 
-    if force or source_repo_changed:
+    if force or source_changed:
         try:
             added_paths, removed_paths, changed_paths, new_entities = update_originals(
-                db_project, now, full_scan=force
+                db_project, now, force=force
             )
         except MissingSourceDirectoryError as e:
             log.error(e)
             return False
 
-        if not db_project.has_single_repo:
-            db_project.source_repository.set_last_synced_revisions()
```

Comment:
It seems like you can keep revert this change now.

And then you don't need any of this:
https://github.com/mozilla/pontoon/pull/1607/files#diff-98331b44e2eb35ae952822657cdd5f6bR173
https://github.com/mozilla/pontoon/pull/1607/files#diff-98331b44e2eb35ae952822657cdd5f6bR221
https://github.com/mozilla/pontoon/pull/1607/files#diff-98331b44e2eb35ae952822657cdd5f6bR236

Reply:
As agreed via chat, it's actually better to set last synced revision at the end of the process to prevent bugs like https://bugzilla.mozilla.org/show_bug.cgi?id=1383271.